pkg/util/idlocker.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 ceph-csi authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"sync"
+)
+
+/*
+IDLock is a per identifier lock with a use counter that retains a number of users of the lock.
+IDLocker is a map of IDLocks holding the IDLocks based on a passed in identifier.
+Typical usage (post creating an IDLocker) is to Lock/Unlock based on identifiers as per the API.
+*/
+type (
+	IDLock struct {
+		mtx      sync.Mutex
+		useCount int
+	}
+
+	IDLocker struct {
+		lMutex sync.Mutex
+		lMap   map[string]*IDLock
+	}
+)
+
+func NewIDLocker() *IDLocker {
```

Comment:
can this a function be renamed as New? IMO the call should not be `idlocker.NewIDLocker()` it should be `idlocker.New()`

Reply:
IMO a map with key as identifier and value as mutex will be enough? the consumer will get the WriteLock during operation and unlock it once it's done.

Reply:
@ShyamsundarR can we use sync.Map inbuilt function?

Reply:
For sync.Map usage, please see point (3) in section "What were the options to improve the locking scheme" in [this](https://github.com/ceph/ceph-csi/pull/443#issuecomment-505129896) comment. I consider this a premature optimization as we are not a CPU bound task in between the locks, hence this should not be required.

Reply:
> can this a function be renamed as New? IMO the call should not be `idlocker.NewIDLocker()` it should be `idlocker.New()`

The call is `NewIDLocker()` and not `idlocker.NewIDLocker()`, IOW we do not have an `idlocker` to start with and are asking for a new instance of the same.

The pattern is inspired by this section in effective go: https://golang.org/doc/effective_go.html#composite_literals

I do not see the problem, am I looking at it wrong?

Reply:
oops sorry got confused with filename and package name. the package name is `util` this should be fine